File: tests/omnia/data/structure/mutable/ArrayQueueTest.java
package omnia.data.structure.mutable;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;

import java.util.Optional;
import omnia.data.structure.List;
import omnia.data.structure.immutable.ImmutableList;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class ArrayQueueTest {

  @Test
  public void init_isEmpty() {
    Queue<?> testSubject = ArrayQueue.create();

    assertFalse(testSubject.isPopulated());
  }

  @Test
  public void init_countIsZero() {
    Queue<?> testSubject = ArrayQueue.create();

    assertEquals(0, testSubject.count());
  }

  @Test(expected = IllegalArgumentException.class)
  public void init_withCapacityZero_didThrowException() {
    ArrayQueue.createWithInitialCapacity(/* capacity= */ 0);
  }

  @Test
  public void enqueue_one_isPopulated() {
    Queue<Object> testSubject = ArrayQueue.create();

    testSubject.enqueue(new Object());

    assertTrue(testSubject.isPopulated());
  }

  @Test
  public void enqueue_one_countIsOne() {
    Queue<Object> testSubject = ArrayQueue.create();

    testSubject.enqueue(new Object());

    assertEquals(1, testSubject.count());
  }

  @Test(expected = NullPointerException.class)
  public void enqueue_null_didThrowException() {
    Queue<Object> testSubject = ArrayQueue.create();

    testSubject.enqueue(null);
  }

  @Test
  public void enqueue_thenDequeue_didReturnObject() {
    Queue<Object> testSubject = ArrayQueue.create();
    Object item = new Object();
    testSubject.enqueue(item);

    Optional<Object> result = testSubject.dequeue();

    assertTrue(result.isPresent());
    assertSame(item, result.get());
  }

  @Test
  public void enqueue_thenDequeue_isEmpty() {
    Queue<Object> testSubject = ArrayQueue.create();
    Object item = new Object();
    testSubject.enqueue(item);

    testSubject.dequeue();

    assertFalse(testSubject.isPopulated());
  }

  @Test
  public void enqueue_thenDequeue_countIsZero() {
    Queue<Object> testSubject = ArrayQueue.create();
    Object item = new Object();
    testSubject.enqueue(item);

    testSubject.dequeue();

    assertEquals(0, testSubject.count());
  }

  @Test
  public void dequeue_whenEmpty_didReturnEmpty() {
    assertFalse(ArrayQueue.create().dequeue().isPresent());
  }

  @Test
  public void enqueue_thenDequeueTwice_didReturnEmpty() {
    Queue<Object> testSubject = ArrayQueue.create();
    testSubject.enqueue(new Object());

    testSubject.dequeue();

    assertFalse(testSubject.dequeue().isPresent());
  }

  @Test
  public void enqueue_fiveHundredTimes_isExpectedSize() {
    List<Integer> data = buildData(500);
    Queue<Integer> testSubject = ArrayQueue.create();

    for (Integer datum : data) {
      testSubject.enqueue(datum);
    }

    assertEquals(500, testSubject.count());
  }

  @Test
  public void enqueue_fiveHundredTimes_thenDequeueFiveHundredTimes_didReturnExpectedItems() {
    List<Integer> data = buildData(500);
    Queue<Integer> testSubject = ArrayQueue.create();

    for (Integer datum : data) {
      testSubject.enqueue(datum);
    }

    for (int i = 0; i < 500; i++) {
      assertEquals(data.itemAt(i), testSubject.dequeue().get());
    }
  }

  @Test
  public void enqueue_fiveHundredTimes_thenDequeueFiveHundredTimes_isEmpty() {
    List<Integer> data = buildData(500);
    Queue<Integer> testSubject = ArrayQueue.create();

    for (Integer datum : data) {
      testSubject.enqueue(datum);
    }

    for (int i = 0; i < 500; i++) {
      testSubject.dequeue().get();
    }
    assertEquals(0, testSubject.count());
    assertFalse(testSubject.isPopulated());
  }

  @Test
  public void enqueue_thenDequeue_fiveHundredTimes_didReturnExpectedItems() {
    List<Integer> data = buildData(500);
    Queue<Integer> testSubject = ArrayQueue.create();

    for (int i = 0; i < 500; i++) {
      Integer item = data.itemAt(i);
      testSubject.enqueue(item);
      assertEquals(item, testSubject.dequeue().get());
    }
  }

  @Test
  public void enqueue_thenDequeue_fiveHundredTimes_isEmpty() {
    List<Integer> data = buildData(500);
    Queue<Integer> testSubject = ArrayQueue.create();

    for (int i = 0; i < 500; i++) {
      Integer item = data.itemAt(i);
      testSubject.enqueue(item);
      testSubject.dequeue();
    }

    assertEquals(0, testSubject.count());
    assertFalse(testSubject.isPopulated());
  }


  @Test
  public void peek_whenEmpty_isEmpty() {
    assertFalse(ArrayQueue.create().peek().isPresent());
  }

  @Test
  public void peek_withOne_didReturnObject() {
    Queue<Integer> testSubject = ArrayQueue.create();
    testSubject.enqueue(132);

    Optional<Integer> datum = testSubject.peek();

    assertTrue(datum.isPresent());
    assertEquals(Integer.valueOf(132), datum.get());
  }

  @Test
  public void peek_withOne_didNotDequeue() {
    Queue<Object> testSubject = ArrayQueue.create();
    testSubject.enqueue(new Object());

    testSubject.peek();

    assertEquals(1, testSubject.count());
    assertTrue(testSubject.isPopulated());
  }

  @Test
  public void peekRepeatedly_withContent_didNotDequeue() {
    Queue<Object> testSubject = ArrayQueue.create();
    for (int i = 0; i < 10; i++) {
      testSubject.enqueue(new Object());
    }

    for (int i = 0; i < 200; i++) {
      testSubject.peek();
    }

    assertEquals(10, testSubject.count());
    assertTrue(testSubject.isPopulated());
  }

  @Test
  public void peekRepeatedly_withContent_didReturnFirst() {
    Queue<Object> testSubject = ArrayQueue.create();
    // First item (which we'll compare with)
    Object expected = new Object();
    testSubject.enqueue(expected);
    // Additional arbitrary objects
    for (int i = 0; i < 10; i++) {
      testSubject.enqueue(new Object());
    }

    for (int i = 0; i < 200; i++) {
      Optional<Object> result = testSubject.peek();
      assertTrue(result.isPresent());
      assertEquals(expected, result.get());
    }
  }

  private static List<Integer> buildData(int count) {
    if (count < 0) {
      throw new IllegalArgumentException(String.format("count must be positive (was %d)", count));
    }
    ImmutableList.Builder<Integer> data = ImmutableList.builder();
    for (int i = 0; i < count; i++) {
      data.add(i);
    }
    return data.build();
  }
}

File: tests/omnia/data/structure/mutable/ArrayQueueTest.kt
package omnia.data.structure.mutable

import com.google.common.truth.Truth.assertThat
import kotlin.test.Test
import kotlin.test.assertFailsWith
import omnia.data.structure.List
import omnia.data.structure.immutable.ImmutableList
import omnia.data.structure.mutable.ArrayQueue.Companion.createWithInitialCapacity

class ArrayQueueTest {

  @Test
  fun init_isEmpty() {
    val testSubject: Queue<*> = ArrayQueue.create<Any>()
    assertThat(testSubject.isPopulated).isFalse()
  }

  @Test
  fun init_countIsZero() {
    val testSubject: Queue<*> = ArrayQueue.create<Any>()
    assertThat(testSubject.count().toLong()).isEqualTo(0)
  }

  @Test
  fun init_withCapacityZero_didThrowException() {
    assertFailsWith(IllegalArgumentException::class) {
      createWithInitialCapacity<Any>(capacity = 0)
    }
  }

  @Test
  fun enqueue_one_isPopulated() {
    val testSubject: Queue<Any> = ArrayQueue.create()
    testSubject.enqueue(Any())
    assertThat(testSubject.isPopulated).isTrue()
  }

  @Test
  fun enqueue_one_countIsOne() {
    val testSubject: Queue<Any> = ArrayQueue.create()
    testSubject.enqueue(Any())
    assertThat(testSubject.count().toLong()).isEqualTo(1)
  }

  @Test
  fun enqueue_null_didThrowException() {
    val testSubject: Queue<Any?> = ArrayQueue.create()
    assertFailsWith(NullPointerException::class) { testSubject.enqueue(null) }
  }

  @Test
  fun enqueue_thenDequeue_didReturnObject() {
    val testSubject: Queue<Any> = ArrayQueue.create()
    val item = Any()
    testSubject.enqueue(item)
    val result = testSubject.dequeue()
    assertThat(result.isPresent).isTrue()
    assertThat(result.get()).isSameInstanceAs(item)
  }

  @Test
  fun enqueue_thenDequeue_isEmpty() {
    val testSubject: Queue<Any> = ArrayQueue.create()
    val item = Any()
    testSubject.enqueue(item)
    testSubject.dequeue()
    assertThat(testSubject.isPopulated).isFalse()
  }

  @Test
  fun enqueue_thenDequeue_countIsZero() {
    val testSubject: Queue<Any> = ArrayQueue.create()
    val item = Any()
    testSubject.enqueue(item)
    testSubject.dequeue()
    assertThat(testSubject.count().toLong()).isEqualTo(0)
  }

  @Test
  fun dequeue_whenEmpty_didReturnEmpty() {
    assertThat(ArrayQueue.create<Any>().dequeue().isPresent).isFalse()
  }

  @Test
  fun enqueue_thenDequeueTwice_didReturnEmpty() {
    val testSubject: Queue<Any> = ArrayQueue.create()
    testSubject.enqueue(Any())
    testSubject.dequeue()
    assertThat(testSubject.dequeue().isPresent).isFalse()
  }

  @Test
  fun enqueue_fiveHundredTimes_isExpectedSize() {
    val data = buildData(500)
    val testSubject: Queue<Int> = ArrayQueue.create()
    for (datum in data) {
      testSubject.enqueue(datum)
    }
    assertThat(testSubject.count().toLong()).isEqualTo(500)
  }

  @Test
  fun enqueue_fiveHundredTimes_thenDequeueFiveHundredTimes_didReturnExpectedItems() {
    val data = buildData(500)
    val testSubject: Queue<Int> = ArrayQueue.create()
    for (datum in data) {
      testSubject.enqueue(datum)
    }
    for (i in 0..499) {
      assertThat(testSubject.dequeue().get()).isEqualTo(data.itemAt(i))
    }
  }

  @Test
  fun enqueue_fiveHundredTimes_thenDequeueFiveHundredTimes_isEmpty() {
    val data = buildData(500)
    val testSubject: Queue<Int> = ArrayQueue.create()
    for (datum in data) {
      testSubject.enqueue(datum)
    }
    for (i in 0..499) {
      testSubject.dequeue().get()
    }
    assertThat(testSubject.count().toLong()).isEqualTo(0)
    assertThat(testSubject.isPopulated).isFalse()
  }

  @Test
  fun enqueue_thenDequeue_fiveHundredTimes_didReturnExpectedItems() {
    val data = buildData(500)
    val testSubject: Queue<Int> = ArrayQueue.create()
    for (i in 0..499) {
      val item = data.itemAt(i)
      testSubject.enqueue(item)
      assertThat(testSubject.dequeue().get()).isEqualTo(item)
    }
  }

  @Test
  fun enqueue_thenDequeue_fiveHundredTimes_isEmpty() {
    val data = buildData(500)
    val testSubject: Queue<Int> = ArrayQueue.create()
    for (i in 0..499) {
      val item = data.itemAt(i)
      testSubject.enqueue(item)
      testSubject.dequeue()
    }
    assertThat(testSubject.count().toLong()).isEqualTo(0)
    assertThat(testSubject.isPopulated).isFalse()
  }

  @Test
  fun peek_whenEmpty_isEmpty() {
    assertThat(ArrayQueue.create<Any>().peek().isPresent).isFalse()
  }

  @Test
  fun peek_withOne_didReturnObject() {
    val testSubject: Queue<Int> = ArrayQueue.create()
    testSubject.enqueue(132)
    val datum = testSubject.peek()
    assertThat(datum.isPresent).isTrue()
    assertThat(datum.get()).isEqualTo(Integer.valueOf(132))
  }

  @Test
  fun peek_withOne_didNotDequeue() {
    val testSubject: Queue<Any> = ArrayQueue.create()
    testSubject.enqueue(Any())
    testSubject.peek()
    assertThat(testSubject.count().toLong()).isEqualTo(1)
    assertThat(testSubject.isPopulated).isTrue()
  }

  @Test
  fun peekRepeatedly_withContent_didNotDequeue() {
    val testSubject: Queue<Any> = ArrayQueue.create()
    for (i in 0..9) {
      testSubject.enqueue(Any())
    }
    for (i in 0..199) {
      testSubject.peek()
    }
    assertThat(testSubject.count().toLong()).isEqualTo(10)
    assertThat(testSubject.isPopulated).isTrue()
  }

  @Test
  fun peekRepeatedly_withContent_didReturnFirst() {
    val testSubject: Queue<Any> = ArrayQueue.create()
    // First item (which we'll compare with)
    val expected = Any()
    testSubject.enqueue(expected)
    // Additional arbitrary objects
    for (i in 0..9) {
      testSubject.enqueue(Any())
    }
    for (i in 0..199) {
      val result = testSubject.peek()
      assertThat(result.isPresent).isTrue()
      assertThat(result.get()).isEqualTo(expected)
    }
  }

  companion object {

    private fun buildData(count: Int): List<Int> {
      require(count >= 0) { String.format("count must be positive (was %d)", count) }
      val data = ImmutableList.builder<Int>()
      for (i in 0 until count) {
        data.add(i)
      }
      return data.build()
    }
  }
}